packages/graphql-hooks/src/types/common-types.ts

```diff
@@ -17,7 +30,18 @@ export type OnErrorFunction<TVariables = any> = ({
   result: Result
 }) => void
 
-export type MiddlewareFunction = () => any
+export type MiddlewareOptions<T> = {
```

Comment:
Using a generic here allows us to specify what extension a middleware may be concerned with, I've converted the existing middleware examples in the repo to Typescript to demonstrate this.